gcp_variant_transforms/libs/variant_partition.py

```diff
@@ -34,59 +44,266 @@
+_REGION_LITERAL_REGEXP = re.compile(r'^(\S+):([0-9,]+)-([0-9,]+)$')
+# A special literal for identifying residual partition's region name.
+_RESIDUAL_REGION_LITERAL = 'residual'
+PARTITION_NOT_FOUND = -1
 
-class VariantPartition(object):
-  """Automatically partition variants based on their reference_name."""
+
+class _ChromosomePartitioner(object):
+  """Assigns partition indices to multiple regions inside a chromosome.
+
+  This class logic is implemented using an interval tree, each region is
+  considered as an interval and will be added to the interval tree. Note all
+  regions must be pairwise disjoint, i.e. no overlapping interval is accepted.
+  """
 
   def __init__(self):
+    # Each instance contains multiple regions of one chromosome.
+    self._interval_tree = intervaltree.IntervalTree()
+
+  def add_region(self, start, end, partition_index):
+    if start < 0:
+      raise ValueError(
+          'Start position on a region cannot be negative: {}'.format(start))
+    if end <= start:
+      raise ValueError('End position must be larger than start position: {} '
+                       'vs {}'.format(end, start))
+    if partition_index < 0:
+      raise ValueError(
+          'Index of a region cannot be negative {}'.format(partition_index))
+    if self._interval_tree.overlaps_range(start, end):
+      raise ValueError(
+          'Cannot add overlapping region {}-{}'.format(start, end))
+    # If everything goes well we add the new region to the interval tree.
+    self._interval_tree.addi(start, end, partition_index)
+
+  def get_partition_index(self, pos=0):
+    """Finds a region that pos falls in it, if none PARTITION_NOT_FOUND."""
+    matched_regions = self._interval_tree.search(pos)
+    # Ensure at most one region is matching to the give position.
+    assert len(matched_regions) <= 1
+    if len(matched_regions) == 1:
+      return next(iter(matched_regions)).data
+    else:
+      return PARTITION_NOT_FOUND
+
+class VariantPartition(object):
+  """Partition variants based on their reference_name and position.
+
+  This class has 2 operating modes:
+    1) No config file is given (config_file_path == None):
+       Automatically partition variants based on their reference_name.
+    2) A config file is given:
+       partitions variants based on input config file.
+  """
+
+  def __init__(self, config_file_path=None):
     if _DEFAULT_NUM_PARTITIONS <= _RESERVED_AUTO_PARTITIONS:
       raise ValueError(
           '_DEFAULT_NUM_PARTITIONS must be > _RESERVED_AUTO_PARTITIONS')
-    self._reference_name_to_partition_map = {}
+    self._num_partitions = _DEFAULT_NUM_PARTITIONS
+    # This variable determines the operation mode auto (default mode) vs config.
+    self._config_file_path_given = False
+    # Residual partition will contain all remaining variants that do not match
+    # to any other partition. -1 means it does not exist.
+    self._residual_partition_index = -1
```

Comment:
Please use the constant instead of `-1` and you can remove the extra comment about the meaning of `-1` :)